slides/04_quality.qmd

```diff
@@ -1,10 +1,655 @@
+
+::: clean-code-rule
+CCR#8: TDD
+:::
+
+![[Photo CC0 by ThisIsEngineering on
+pexels.com]{.copyright}](thumbnails/testing.jpg){fig-align="center"}
+
+## Verification vs Validation
+
+::: columns
+::: {.column width="50%"}
+**Verification**:\
+Are we building the product right?
+
+-   The product meets a set of specified requirements
+-   Activities: reviews, inspections, and testing
+-   Goal: ensure that the product has been designed and implemented correctly
+:::
+
+::: {.column width="50%"}
+**Validation**:\
+Are we building the right product?
+
+-   Evaluate the product during or at the end of the development process
+-   Show that the product meet the specified requirements
+-   Goal: confirm that the product is fit for its intended purpose\
+    $\rightarrow$ Refers back to the user's needs
+:::
+:::
+
+## What are Unit Tests?
+
+-   Automated tests
+-   Record the expected output of a function using code
+-   Check individual units of code (functions or methods) for correctness
+-   Typically written by developers as part of the software development process
+-   Typically run automatically as part of a CI/CD pipeline
+-   Can be run before committing code changes
+
+## Why are unit tests important?
+
+::: clean-code-rule
+CCR#8: TDD
+:::
+
+-   Help to ensure that individual units of code are working correctly
+-   Ensure that changes to the codebase do not break existing functionality
+-   Express the desired behavior in a way that a human can understand
+-   Help to identify and fix bugs early on in the development process
+-   Save time and resources by catching issues before they become more difficult
+    and costly to fix
+
+::: info-block
+::: columns
+::: {.column width="20%"}
+[{{< fa circle-info size=3x >}}]{.yellow}
+:::
+
+::: {.column width="80%"}
+Unit tests help to increase the reliability and maintainability of the code
+:::
+:::
+:::
+
+## What other important test types exist?
+
+-   **Integration Testing**: Test if different functions or scripts work
+    together as expected
+-   **Performance Testing**: Analyze the performance of the implemented
+    functions and check whether they meet the requirements (UR/UX)
+-   **Snapshot Testing**: Record the results in a separate human-readable file
+    and compare it to the output during the test; useful for large/complex
+    outputs and binary formats like plots
+
+# Testing in Practice
+
+![[Photo CC0 by cottonbro studio on
+pexels.com]{.copyright}](thumbnails/coding_cottonbro-studio.jpg){fig-align="center"}
+
+## How to realize testing with R?
+
+R package [testthat](https://cran.r-project.org/package=testthat)
+
+-   Popular testing framework for R that is easy to learn and use
+-   Unit testing, integration testing, and [snapshot
+    testing](https://cran.r-project.org/web/packages/testthat/vignettes/snapshotting.html)
+    supported
+-   Also performance testing, e.g., with help of
+    [microbenchmark](https://cran.r-project.org/package=microbenchmark) package
+
```

Comment:
I think one pr twp more slides here to show how this looks like in R will be important and useful, you can borrow shamelessly from https://rconsortium.github.io/asa-biop-swe-wg/slides/unit_testing_basic.html (source is in the repo)

Reply:
Good idea 👍 The unit testing basic slides are realy good!